src/connect.rs

```diff
@@ -41,29 +44,82 @@ pub(crate) mod encryption {
         match mode {
             Mode::Plain => Ok(StreamSwitcher::Plain(socket)),
             Mode::Tls => {
-                let try_connector = tls_connector.map_or_else(|| TlsConnector::new(), |c| Ok(c));
-                let connector = try_connector.map_err(Error::Tls)?;
+                let try_connector = tls_connector.map_or_else(TlsConnector::new, Ok);
+                let connector = try_connector.map_err(TlsError::Native)?;
                 let stream = TokioTlsConnector::from(connector);
                 let connected = stream.connect(&domain, socket).await;
                 match connected {
-                    Err(e) => Err(Error::Tls(e)),
+                    Err(e) => Err(Error::Tls(e.into())),
                     Ok(s) => Ok(StreamSwitcher::Tls(s)),
                 }
             }
         }
     }
 }
 
-#[cfg(feature = "tls")]
-pub use self::encryption::MaybeTlsStream;
+#[cfg(all(feature = "rustls-tls", not(feature = "native-tls")))]
+pub(crate) mod encryption {
+    pub use rustls::ClientConfig;
+    use tokio_rustls::{webpki::DNSNameRef, TlsConnector as TokioTlsConnector, TlsStream};
+
+    use std::sync::Arc;
+    use tokio::io::{AsyncRead, AsyncWrite};
+
+    use tungstenite::{error::TlsError, stream::Mode, Error};
+
+    use crate::stream::Stream as StreamSwitcher;
+
+    /// A stream that might be protected with TLS.
+    pub type MaybeTlsStream<S> = StreamSwitcher<S, TlsStream<S>>;
+
+    pub type AutoStream<S> = MaybeTlsStream<S>;
+
+    /// A TLS connector that can be used when establishing TLS connections.
+    pub type TlsConnector = Arc<ClientConfig>;
```

Comment:
There is a similar concern as in `tungstenite`: `MaybeTlsStream` and `TlsConnector` end up in public API sufrace of tokio-tungstenite. One crate may pull `tokio-tungstenite/rustls-tls`, and if another crate pulls `tokio-tungstenite/native-tls`, then the public API changes (possibly breaking the first crate).

The (reasonable) requirement that feature flags should be additive is really making our lives harder here.

We should probably try to somehow encapsulate this. Brainstorm: speaking about `MaybeTlsStream`, it could be also possible to implement the "dispatch" on Stream(Switcher), possibly making it `#[non_exhaustive]`. The question is whether such an encapsulation of the underlying stream would be limiting to our users - but I see that fields of Stream(Switch) are already private.

Reply:
On seconds thought, maybe not that ready for review/merge yet 😅 

So after looking through the `MaybeTlsStream` and `TlsConnector` usages again I figured that both of them are quite troublesome.

- `MaybeTlsStream` can be extracted out of the `WebSocketStream` with the `get_ref` and `get_mut` functions. That means there is a chance that users may rely on the inner type being from `rustls` or `native-tls`.
- `TlsConnector` can be passed to `client_async_tls_with_config` which allows to customize the connector beforehand.

Encapsulating any of these two seems to either create a weird API or reduce customization options for the user. 

For example encapsulating `MaybeTlsStream` would either mean we hide the inner stream giving the user no access to its inner stream anymore. Or we use an enum but the users would need to match on it every time which is weird if you only have one TLS flag enabled (which is the case most of the time).

If we turn `TlsConnector` into an enum we would make it possible for users to pass the wrong connector to the function which would break things anyways. Say for example we pass a `rustls` connector but if both features are enabled it would use `native-tls` and the connector would not match so we would have to return an error anyways.

There is some way for sure to solve this nicely but currently I didn't find that way yet. I totally agree that the current implementation would break things, just not sure how to solve this well.

I will think about it a few days and see whether I can come up with something better.

Reply:
Good points that the enum-based solution would be rather subpar.

Food for thought: we can side-step the problem by (conditionally) exposing 2 different streams: MaybeNativeTlsStream and MaybeRustTlsStream (or some shorter variant, or expose the generic parameter). We can help our users by making their API "textually the same", but not actually interchangeable in rustc sense. An analogy to std's BTreeMap vs. HashMap.

Optionally, we could provide some wrapper type on top of that for users that don't care about the particular implementation (majority?). Such a wrapper would come with the limitation that you cannot access the inner stream.

That's best I can think of right now. Feel encouraged to take your time and come up with own ideas - there's no need to hurry here.

Reply:
I made some rather big changes to the PR today, trying to solve this problem. Currently I'm quite happy with the result and found a solution where you can even control the used implemention with the `TlsConnector`.

I will do some more changes on it but if you want you can already have a look.

I'm thinking about whether the new `TlsStream` enum could be merged with the `MaybeTlsStream` 🤔

Reply:
Maybe good to move the encryption module into its own file, but I would have to rename `connect.rs` to `mod.rs` and that makes the changes really hard to diff and review.

So for the sake of the review I would say better to do that as a separate PR afterwards.